### PR TITLE
Add LLVMGPU transform to rewrite scf.foreach_thread to gpu.thread + translation_info attr.

### PR DESCRIPTION
**Pull request by nicolasvasilache**
An `scf.foreach_thread` is mapped 1-1 to a threadblock for now and sets the threadblock launch config.
It is the responsibility of the user of the pattern to make sure multiple such `scf.foreach_thread` would be compatible within the same kernel.
Enhancements to this basic mode will come in the future on a per-need basis, should be pretty easy to extend.

### CODE REVIEW

**Comment by benvanik** [comment]
(Please use more descriptive PR titles when pushing to the main repo - everyone sees these :)

**Comment by benvanik** [comment]
Yep - everything in the repo will show on discord and on github dashboard for anyone subscribed to the repo:
image
image
(this showed up, but got renamed when you mentioned me)

IOW, all shame is public when pushed to the repo, so best to treat everything you do as fully visible unless it's in your fork :)

**Comment by nicolasvasilache** [comment]
I see, thanks @benvanik !

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformDialectExtensions/TransformDialectLLVMGPUExtensions.cpp [hunk 1]
```diff
@@ -27,5 +35,118 @@ void mlir::iree_compiler::registerTransformDialectLLVMGPUExtension(
...
+  rewriter.eraseOp(foreachThreadOp);
+
+  return threadCount;
+}
+
+//===---------------------------------------------------------------------===//
+// IREE-specific LLVMGPU transformations.
+//===---------------------------------------------------------------------===//
+
+// TODO: if the number of threads was wired like the workgroup_count, we could
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformDialectExtensions/TransformDialectLLVMGPUExtensions.cpp:100]: Maybe I am mistaken, but I see direct mapping to threadids here?

Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformDialectExtensions/TransformDialectLLVMGPUExtensions.cpp:100]: not sure how to parse this, is it related to the TODO ?
The TODO says we could have non-static dimensions too, although we prob often want the `tidx` to be static and a multiple of warp size (but I've seen many cases of it not being the best thing to do).

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformDialectExtensions/TransformDialectLLVMGPUExtensions.cpp:100]: I am trying to understand how the body of the `for_each` gets mapped to the hardware. Does the body map to a thread or a thread block? Seems like it is directly mapping to a thread?

Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformDialectExtensions/TransformDialectLLVMGPUExtensions.cpp:100]: Yes, a foreach is mapped 1-1 to a threadblock for now and sets the threadblock.
It is the responsibility of the user of the pattern to make sure multiple such foreach_thread would be compatible within the same kernel.
Enhancements to this basic mode will come in the future on a per-need basis, should be pretty easy to extend.

**Review by ThomasRaoux — APPROVED** [review]

**Pull request merged**